Add Statement and Database\Query::all()

othercorey opened the pull request:
This has several key changes:

* Removes support for callbacks in statements.
* Remove nested statement design in favor of single implementation.
* Move field type conversion directly into statement instead of using callbacks.
* Add Database\Query::all() which returns a result set for SELECT statements.

Users would be required to call `all()` for result set decorators to run. 

This enforces result set modification is done outside of the Statement which should be a straight-forward wrapper around `PDOStatement`.

Key issues;

* ~`Database\Query` owns result decorators which used to run via callbacks, but now must be wrapped in a result set separate from `execute()`.~
* ~`Database\Query::all()` returns `iterable` instead of `array` to allow `Datasource\QueryTrait::all()` to override it with `ResultSetInterface`.~
* `QueryTrait` declares `$_results` but it's already declared in `Database\Query`. This is because `ORM\Query` is implemented by a Datasource trait.
* `SqliteStatement` is still guessing what type of query was run to perform `SELECT CHANGES()`.

othercorey committed:
Add Database\Statement and Database\Query::all().

Statements no longer use a nested decorator pattern. Callbacks were removed.

othercorey commented on src/ORM/Query.php:
```diff
@@ -1126,7 +1125,8 @@ protected function _execute(): iterable
         }
 
-        $results = $this->execute()->fetchAll(StatementInterface::FETCH_TYPE_ASSOC);
+        /** @var array $results */
```
This is a hack since we know Database\Query actually uses an array.

ADmad replied:
We can look into changing the arguments which accept this resultset to `iterable` instead of `array`.

othercorey replied:
Updated to convert to `array` before passing to `loadExternal()`. Will work on results sets separately.

othercorey commented on src/Database/Query.php:
```diff
@@ -237,15 +242,40 @@ public function getConnection(): Connection
 
-        return $this->_iterator;
+        return $this->_statement;
+    }
```
This should throw an exception if result decorators were added, but I'm currently calling `execute()` from `all()`.

markstory replied:
Could we raise an error if `_type !== Query::TYPE_SELECT && $this->_resultDecorators`?

othercorey replied:
I call execute() from all(). I'd have to move this to a protected method.

ADmad commented on src/Database/Statement.php:
```diff
@@ -0,0 +1,370 @@
+use PDOStatement;
+use Psr\Log\LoggerInterface;
+
+class Statement
```
Why did you remove the interface?

othercorey replied:
Without the nested design, there is no need for an interface. We expect all instances to extend `Statement`.

ADmad replied:
Hmm.. okay. `DriverInterface` is quite redundant too, no one's going to implement their own implementation of it. `Driver` class will be always used.

markstory replied:
If we have no scenario where we expect users to swap in their own statements then we could do without the interfaces. It does force us to use inheritance in our other statements instead of decoration as well.

othercorey replied:
I updated the design to support inheritance in driver-specific statements better. Notice how `SqlserverStatement` doesn't have to duplicate the bind logic anymore.

othercorey replied:
Re-introduced StatementInterface.

ADmad commented on src/Database/Driver.php:
```diff
@@ -198,12 +202,18 @@ abstract public function enabled(): bool;
 
-        return new PDOStatement($statement, $this);
+        $typeMap = null;
+        if ($query instanceof Query && $query->isResultsCastingEnabled() && $query->type() === 'select') {
```
Would be nice to have constants for query types instead of using strings.

othercorey replied:
I can add those back.

othercorey replied:
Whoops, I misread this. Will have to add new constants.

othercorey replied:
Added TYPE_SELECT constant to Query.

ADmad commented on src/Database/Driver/Sqlserver.php:
```diff
@@ -212,7 +209,12 @@ public function prepare(Query|string $query): StatementInterface
+            $typeMap = $query->getSelectTypeMap();
+        }
+
+        return new SqlserverStatement($this, $statement, $typeMap);
```
The `STATEMENT_CLASS` constant should be used.

othercorey replied:
Fixed.

ADmad commented on src/Database/Query.php:
```diff
@@ -2395,7 +2404,7 @@ function ($errno, $errstr): void {
             'defaultTypes' => $this->getDefaultTypes(),
             'decorators' => count($this->_resultDecorators),
-            'executed' => $this->_iterator ? true : false,
+            'executed' => $this->_statement !== null ? true : false,
```
```suggestion
            'executed' => (bool)$this->_statement,
```

ADmad commented on src/Database/Statement.php:
```diff
@@ -0,0 +1,370 @@
+     */
+    public function execute(?array $params = null): bool
+    {
+        $loggedQuery = new LoggedQuery();
```
Shouldn't the `LoggedQuery` instance be created only if logging is enabled?

othercorey replied:
I would like to clean up this function, but I didn't figure anything out yet. I do want to avoid a bunch of if checks and temp variables. The requirement to capture and log the error makes things a little clunky.

ADmad replied:
Yes, the currently code is quite clunky.

markstory replied:
Could we take a similar approach to how we do error logging where the metadata (duration, row counts) is sent to a logger that then decides on whether or not to log the query? That might be a breaking change but again I suspect there are very few database logging implementations in the wild so the impact would be low.

ADmad replied:
Not sure if that would be the right approach. Logging is an optional feature. So if it's not enabled then a logger instance wouldn't/shouldn't exist.

othercorey replied:
Cleaned up logging in execute(). Only runs if logger is set.

ADmad commented on src/Database/Statement.php:
```diff
@@ -0,0 +1,370 @@
+            return $mode;
+        }
+
+        static $MODES = ['assoc' => PDO::FETCH_ASSOC, 'num' => PDO::FETCH_NUM, 'obj' => PDO::FETCH_OBJ];
```
Why not use a protected static class property?

othercorey replied:
I did originally. Tried it this way to see if it was cleaner.

othercorey replied:
Fixed.

ADmad commented on src/Database/Statement.php:
```diff
@@ -0,0 +1,370 @@
+     */
+    public function rowCount(): int
+    {
+        if ($this->typeConverter !== null) {
```
So this will effectively prevent use of `rowCount()` for SELECT queries even for dbs that support it isn't it? Not sure if we should do that. It could still be useful for debugging.

othercorey replied:
I don't believe it's ever supported that way, but we can do that.

ADmad replied:
Well `rowCount()` does work for all (most) dbs besides Sqlite even for SELECT and is shown in DebugKit currently.

othercorey replied:
Fixed.

othercorey commented:
I removed ``lastInsertId()` because of the support it implies.

If the driver does not return inserted IDs as rows, then it calls `PDO::lastInsertId()` which does not operate on a specific statement or column and for some drivers not on a specific table (postgresqL

However, for drivers like MySQL, having to find the driver each time you want the auto-increment id, that might be tedious. This is hidden from the user for ORM inserts.

ADmad commented on src/Datasource/FixtureInterface.php:
```diff
@@ -16,7 +16,7 @@
 namespace Cake\Datasource;
 
-use Cake\Database\StatementInterface;
+use Cake\Database\Statement;
```
This is a separate issue but datasource package shouldn't have a dependency on a database package class :disappointed:

ADmad commented on src/ORM/Table.php:
```diff
@@ -2105,9 +2105,14 @@ protected function _insert(EntityInterface $entity, array $data): EntityInterfac
             $entity->set($filteredKeys, ['guard' => false]);
             $schema = $this->getSchema();
             $driver = $this->getConnection()->getDriver();
+            $returning = $statement->fetchAssoc();
```
This is better than using `PDO::lastInsertId()`?

othercorey replied:
For drivers that support RETURNING and similar syntax, the query returns a row with the generated IDs. PDO::lastInsertId() either returns the last ID for a table or for a sequence. For example, you can't call lastInsertId() for a table on postgresql. You must fetch the row retuned from INSERT.

othercorey replied:
Reverted.

markstory commented on src/Database/Connection.php:
```diff
@@ -448,9 +446,9 @@ public function update(string $table, array $values, array $conditions = [], arr
      * @param array $conditions conditions to be set for delete statement
      * @param array $types list of associative array containing the types to be used for casting
-     * @return \Cake\Database\StatementInterface
+     * @return \Cake\Database\Statement
```
Why don't we want to return an interface here?

othercorey replied:
There's no reason to support an Interface now that we don't use the nested decorator pattern. We expect a `Statement`.

othercorey replied:
Re-introduced StatementInterface.

markstory commented on src/Database/Query.php:
```diff
@@ -149,11 +149,16 @@ class Query implements ExpressionInterface, IteratorAggregate, Stringable
+     * @var iterable|null
      */
-    protected ?StatementInterface $_iterator = null;
+    protected ?iterable $_results = null;
```
If we're adding new properties could we skip the `_` prefix?

othercorey replied:
This is caused by `QueryTrait`. ORM\Query doesn't actually implement its own features. It uses `QueryTrait` which defines `$_results`. 

It's awkward to have a trait overrides the base class, but it seemed better to line it up.

markstory commented on src/Database/Query.php:
```diff
@@ -237,15 +242,40 @@ public function getConnection(): Connection
+    }
+
+    /**
+     * Executes query and returns entire result set.
```
This reads like the output will be a ResultSet.
```suggestion
     * Executes query and returns the decorated results.
```

othercorey commented on src/Database/Connection.php:
```diff
@@ -280,9 +279,8 @@ public function prepare(Query|string $query): StatementInterface
-
             if ($this->_logQueries) {
-                $statement = $this->_newLogger($statement);
+                $statement->setLogger($this->getLogger());
```
Should we make `setLogger()` part of `StatementInterface`? Otherwise, we have to type that this is a `Statement` instance.

ADmad replied:
Could we instead come up with a solution which doesn't require passing the logger into the statement? The driver could do the logging after it executes the statement.

othercorey replied:
Added to StatementInterface for now. Will revisit separately.

markstory replied:
Since statements already have access to the driver, could we get the logger from the driver? If logging is disabled we could use a NullObject to do a no-op.

othercorey commented:
Update:

* Re-introduced `StatementInterface`. We can revisit that later.
* Moved `Statement` to `Database\Statement\Statement`
* Changed `ORM\Query::all()` to convert iterable to array before passing to `EagerLoader`
* Added `Query::TYPE_*` constants to replace `'select'`
* Added Statement::queryString()
* Cleaned up query logging
* Added `StatementTest` which ports the test from `LoggingStatementTest`

ADmad commented on src/Database/Statement/Statement.php:
```diff
@@ -0,0 +1,328 @@
+    /**
+     * @var \PDOStatement
+     */
+    protected PDOStatement $statement;
```
I would rename this to `$pdoStatement` for better readability.

othercorey replied:
I can do that.

markstory commented on src/Database/Statement/Statement.php:
```diff
@@ -0,0 +1,328 @@
+     */
+    public function setLogger(LoggerInterface $logger)
+    {
+        $this->logger = $logger;
```
Instead of setting the logger on each statement, could we get the logger from the driver? It would require pushing the logger down from `Connection` into the `Driver` but it would remove some complexity here.

If you'd like I can take a swing at doing that change.

othercorey replied:
What is the difference between Connection and Driver long-term? Do we want Driver to maintain state?

markstory replied:
Long term I don't know. The entire Datasources namespace is a bit suspect and I'm not sure it is justified if we are going to focus on the just the ORM and make plugins like Elasticsearch totally separate but 'ORM like'.

othercorey committed:
Clean up Query::getIterator() and add test

othercorey committed:
Merge pull request #16332 from cakephp/query-iterator

Clean up Query::getIterator() and add test
